### PR TITLE
[Bitcoin/Rust] Add support for creating Ordinal NFT inscriptions

### PR DESCRIPTION
**Pull request by lamafab**
This PR allows you to create Ordinal NFT inscriptions, example: https://www.blockchain.com/explorer/transactions/btc/173f8350b722243d44cc8db5584de76b432eb6d0888d9e66e662db51584f44ac

Generally ready, two things to consider:

1. The `TransactionVariant` protobuf enum now has the variants:

```proto
enum TransactionVariant {
    P2PKH = 0;
    P2WPKH = 1;
    P2TRKEYPATH = 2;
    BRC20TRANSFER = 3;
    NFTINSCRIPTION = 4;
}
```

The last two can technically be combined, I propose calling it `P2ScriptPath` (`P2TRKEYPATH` should be lowercased, too). But that's a breaking change afaik.

EDIT: We're keeping it like this now, we're planning on writing a new interface as part of a larger task anyway.

2. ~Would be nice if someone could quickly write `Script::buildNftInscription` in `src/Bitcoin.cpp` in order to expose it. Note that one has to pass on the `MimeType` from `src/nft.rs`.~ Done

EDIT: Additionally, added tests for C++ and Swift, Kotlin is WIP... having some build/dependency error (?). Will change this from Draft to Open, let the CI run.

### CODE REVIEW

File: include/TrustWalletCore/TWOrdMimeType.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+#pragma once
+
+#include "TWBase.h"
+
+TW_EXTERN_C_BEGIN
+
+/// The Bitcoin inscription MIME type.
+TW_EXPORT_ENUM(uint32_t)
+enum TWOrdMimeType {
+    TWOrdMimeTypeApplicationJson = 1,
```
Comment (satoshiotomakan) [inline include/TrustWalletCore/TWOrdMimeType.h:16]: Will the Bitcoin network reject a BRC20 transaction if the mime type doesn't match this list?
Probably, we could allow the user to pass a mime type as a string? 🤔

File: rust/tw_bitcoin/src/ffi/mod.rs [hunk 2]
```diff
@@ -155,9 +155,13 @@ pub(crate) fn taproot_build_and_sign_transaction(proto: SigningInput) -> Result<
             TrVariant::P2TRKEYPATH => {
                 TxOutputP2TRKeyPath::new_with_script(satoshis, script_buf).into()
             },
+            // TODO: We could combine the next two variants into a single one.
```
Comment (satoshiotomakan) [inline rust/tw_bitcoin/src/ffi/mod.rs:158]: TODO

File: rust/tw_bitcoin/src/ffi/scripts.rs [hunk 3]
```diff
@@ -142,3 +143,47 @@ pub unsafe extern "C" fn tw_build_brc20_inscribe_transfer(
...
+    // Convert Recipient.
+    let slice = try_or_else!(
+        CByteArrayRef::new(pubkey, pubkey_len).as_slice(),
+        CByteArray::null
+    );
+
+    let recipient = try_or_else!(Recipient::<PublicKey>::from_slice(slice), CByteArray::null);
+
+    // Inscribe NFT data.
+    let nft = NftInscription::new(mime_type, data, recipient).expect("TODO");
```
Comment (satoshiotomakan) [inline rust/tw_bitcoin/src/ffi/scripts.rs:174]: TODO

File: src/Bitcoin/Script.cpp [hunk 4]
```diff
@@ -531,7 +531,22 @@ Script Script::lockScriptForAddress(const std::string& string, enum TWCoinType c
...
-    Rust::CByteArrayWrapper res = TW::Rust::tw_build_brc20_inscribe_transfer(tickerBytes.data(), amount, 0, publicKey.data(), publicKey.size());
+    Rust::CByteArrayWrapper res = TW::Rust::tw_build_brc20_transfer_inscription(tickerBytes.data(), amount, 0, publicKey.data(), publicKey.size());
+    auto result = res.data;
+    out.ParseFromArray(result.data(), static_cast<int>(result.size()));
+    return out;
+}
+
+Proto::TransactionOutput Script::buildNftInscription(enum TWOrdMimeType mimeType, const Data& payload, const Data& publicKey) {
+    TW::Bitcoin::Proto::TransactionOutput out;
+    Rust::CByteArrayWrapper res = TW::Rust::tw_build_nft_inscription(
```
Comment (satoshiotomakan) [inline src/Bitcoin/Script.cpp:542]: `tw_build_nft_inscription` is ambiguous. How about `tw_build_brc20_nft_inscription`?

File: swift/Tests/Blockchains/BitcoinTests.swift [hunk 5]
```diff
@@ -6,6 +6,7 @@
 
 import XCTest
 import WalletCore
+import Foundation
```
Comment (satoshiotomakan) [inline swift/Tests/Blockchains/BitcoinTests.swift:9]: Is this import needed?

**Review by satoshiotomakan — COMMENTED** [review]
Good job! Just a few questions/suggestions/reminders

File: src/Bitcoin/Script.cpp (hunk 4 above)
Comment (lamafab) [reply src/Bitcoin/Script.cpp:542]: maybe `tw_build_ordinal_nft_inscription`, it's not based on BRC20

File: include/TrustWalletCore/TWOrdMimeType.h (hunk 1 above)
Comment (lamafab) [reply include/TrustWalletCore/TWOrdMimeType.h:16]: Yes, we could add a `Custom(String)` variant

Comment (lamafab) [reply include/TrustWalletCore/TWOrdMimeType.h:16]: Although this makes the FFI interface a little more complex... Maybe just pass a string directly, as you said. Will update.

File: src/Bitcoin/Script.cpp (hunk 4 above)
Comment (satoshiotomakan) [reply src/Bitcoin/Script.cpp:542]: It's related to Bitcoin, so probably `tw_bitcoin_nft_inscription` then?

File: rust/tw_bitcoin/src/ffi/scripts.rs [hunk 6]
```diff
@@ -146,15 +146,20 @@ pub unsafe extern "C" fn tw_build_brc20_transfer_inscription(
...
     data: *const u8,
     data_len: usize,
     satoshis: i64,
     pubkey: *const u8,
     pubkey_len: usize,
 ) -> CByteArray {
-    let mime_type = try_or_else!(MimeType::from_raw(mime_type as usize), CByteArray::null);
+    // Convert mimeType.
+    let mime_type = try_or_else!(
+        CByteArrayRef::new(mime_type, mime_type_len).as_slice(),
```
Comment (satoshiotomakan) [inline rust/tw_bitcoin/src/ffi/scripts.rs:160]: Should `mime_type` be a string?
An example
https://github.com/trustwallet/wallet-core/blob/f266567e913a40bfee80b8ecdd4b74785cea2b32/rust/tw_encoding/src/ffi.rs#L97-L100

But I plan to implement `TWString` same way as in C++

File: samples/kmp/iosApp/Podfile.lock [hunk 7]
```diff
@@ -1,12 +1,12 @@
...
-  - TrustWalletCore (2.9.8):
-    - TrustWalletCore/Core (= 2.9.8)
-  - TrustWalletCore/Core (2.9.8):
+  - SwiftProtobuf (1.22.0)
+  - TrustWalletCore (3.2.5):
+    - TrustWalletCore/Core (= 3.2.5)
+  - TrustWalletCore/Core (3.2.5):
     - TrustWalletCore/Types
-  - TrustWalletCore/Types (2.9.8):
+  - TrustWalletCore/Types (3.2.5):
```
Comment (satoshiotomakan) [inline samples/kmp/iosApp/Podfile.lock:9]: Could you please revert these changes in this file? 😅

File: samples/osx/cocoapods/Podfile.lock [hunk 8]
```diff
@@ -1,11 +1,11 @@
 PODS:
   - BigInt (5.2.0)
-  - SwiftProtobuf (1.19.0)
```
Comment (satoshiotomakan) [inline samples/osx/cocoapods/Podfile.lock:3]: Here too

File: swift/Example/Podfile.lock [hunk 9]
```diff
@@ -1,5 +1,5 @@
 PODS:
-  - SwiftProtobuf (1.16.0)
```
Comment (satoshiotomakan) [inline swift/Example/Podfile.lock:2]: Here too

File: swift/Podfile.lock [hunk 10]
```diff
@@ -1,5 +1,5 @@
 PODS:
-  - SwiftProtobuf (1.11.3)
```
Comment (satoshiotomakan) [inline swift/Podfile.lock]: Here too

**Review by satoshiotomakan — APPROVED** [review]
LGTM 🔥

**Review by Milerius — APPROVED** [review]
